src/box/txn.c

```diff
@@ -768,7 +768,7 @@ txn_complete_fail(struct txn *txn)
 	assert(txn->signature != TXN_SIGNATURE_UNKNOWN);
 	assert(in_txn() == txn);
 	if (txn->limbo_entry != NULL) {
-		assert(txn_has_flag(txn, TXN_WAIT_SYNC));
+		assert(txn_has_flag(txn, TXN_IS_SYNC));
```

Comment:
This is just an auxiliary flag so we can detect transactions that go to the limbo with one `txn_has_flag` check. Previously, I tried to introduce a `txn_must_be_in_limbo` helper for this without introducing a new flag, but I received the followig feedback about it: https://github.com/tarantool/tarantool/pull/10249#pullrequestreview-2194370609

Reply:
@sergepetrenko lets discuss. I don't feel confident about introduction of new flags which are cheap to deduct from the existing states of other things. We really only have 2 transaction types which are not deductible from anything (not even from the changed spaces, because we have `is_sync` for begin and commit funcs): synchronous and asynchronous. Everything else is deductible. And we can do that via the flags `WAIT_SYNC` and `WAIT_ACK`. Strictly speaking, I am not even sure we need both of them.

But now we are having `WAIT_SYNC`, `WAIT_ACK`, `IS_SYNC`, `EARLY_ACK`. Can anybody without looking at the code correctly tell the difference between `WAIT_SYNC` and `IS_SYNC`? Or `WAIT_ACK` and `EARLY_ACK`. These make impression like we have 16 transaction types (all combinations of these flags), but we only have 2. The rest fully depend on when the txn was created. In one moment it would have `EARLY_ACK`, but created a bit later it wouldn't?

What is so bad about introducing a few check-functions which would check txn behaviour depending on its synchronousness and limbo state? We could even add a new enum for this. But putting this into the flags feel super confusing 😵‍💫. Am I looking at it from a wrong angle?

Reply:
@Gerold103, thanks for bringing this up!
Your comment's got me thinking big time.

It was my misconception that all the new flags would be needed for the replica to behave correctly and distinguish between transaction types. And I thought we needed at least 3 distinct flags to make it all work. That's why I suggested to reshuffle the flags to make each of them responsible for a particular property (`txn_should_go_to_limbo`, `txn_needs_ack`, etc).
Actually, we don't need all these flags on a replica. Indeed, all the "transaction types" can be deduced quite easily by the replica.

For example: the fact that txn should go to limbo may be calculated like `TXN_WAIT_ACK || limbo_is_not_empty || replication_split_brain_handling_mode != 'none'`.

`TXN_WAIT_SYNC` is obviously replaced by a check for limbo being empty.

Given your other comment about storing completed transactions in the limbo, `TXN_EARLY_ACK == txn_is_committed`

Actually, we only need a single flag, `TXN_WAIT_ACK`, which marks truly synchronous transactions. But unfortunately we already have an extra one in the protocol - `TXN_WAIT_SYNC`, so it'll have to stay for backward compatibility.

Long story short, let's get rid of `TXN_EARLY_ACK` completely. And replace a check for it with a check that txn is committed.

P.S. Also need to make sure that all this reasoning stays correct in light of the synchronous promote.

Reply:
> Indeed, all the "transaction types" can be deduced quite easily by the replica.

How can the replica deduce whether it should keep the asynchronous transaction in the limbo after a preceding synchronous one gets committed?

Reply:
> How can the replica deduce whether it should keep the asynchronous transaction in the limbo after a preceding synchronous one gets committed, if we want it to follow the master's behaviour?

After giving it some thought, I think we can let the replica do whatever it is told to do by its `replication_split_brain_handling_mode`. So it can behave differently. It can drop the async transactions from the limbo, if it doesn't have the `split_brain_handling_mode` turned on. And it might get a split-brain then, but that's expected. That's what the option is meant for.

Have I forgot about some reasoning here? Should we really make the replica follow master's behavior regardless of local `split_brain_handling_mode` setting?

Reply:
> Given your other comment about storing completed transactions in the limbo, `TXN_EARLY_ACK == txn_is_committed`.
> Long story short, let's get rid of TXN_EARLY_ACK completely. And replace a check for it with a check that txn is committed.

We need to determine that a transaction should be acked to the user early after we return from the WAL in `txn_on_journal_write`. We cannot distinguish such transactions from synchronous ones, since we need to set the `TXN_WAIT_ACK` flag early, since we need to set the `TXN_WAIT_ACK` flag for replication to work properly.